### ISSUE TITLE
Add missing `predict_proba` and `predict_log_proba` generic methods when applicable

### ISSUE THREAD

**Issue by terrytangyuan** [h2oai/h2o4gpu#433]
Add them as generic methods for now. We can also consider a single `predict()` function with argument `type = c("probability", "raw", "log_probability")` that is more old R fashioned. @ledell @navdeep-G Any preference here?

**Comment by ledell** [comment]
I think we can skip the `predict_proba()` and `predict_log_proba()` functions in the R API (at least directly exposed to the user).  

We should definitely support `predict()` and so we'd just have to decide on what we'd like to output from that.  

1. We could follow the H2O convention of returning:
- k+1 columns for k-class classification (first column is predicted class, additional columns are predicted values for each class).
- single column for regression (just the raw predicted value).

2. Or use the `type` argument to produce different results. 
- in R, usually `type = "response"` gives raw predicted values (or classes) and `"prob"` gives a matrix of "probabiliities" though unlike sklearn's `predict_proba()` which uses Platt scaling, these are not really probabilities, so the name is not great.  So one option is to allow types `"response"` and `"prob"`.
- One concern I have about using `predict_proba()` underneath to produce probabilities is that it only works for sklearn models which is probably not what people will be using 95% of the time, and also requires [much more computation](https://stackoverflow.com/a/15112563/5451344).  So my opinion would be not enable a `type` option which attempts to call either `predict_proba()` `predict_log_proba()` underneath.  

Here's a few examples from R for demonstration:

Classification:
```
> library(randomForest)
> rf <- randomForest(x = iris[,1:4], y = iris[,5])
> head(predict(rf, iris, type = "response"))
     1      2      3      4      5      6 
setosa setosa setosa setosa setosa setosa 
Levels: setosa versicolor virginica
> head(predict(rf, iris, type = "prob"))
  setosa versicolor virginica
1  1.000      0.000         0
2  0.998      0.002         0
3  1.000      0.000         0
4  1.000      0.000         0
5  1.000      0.000         0
6  1.000      0.000         0
```

Regression:
```
> rf <- randomForest(x = iris[,1:3], y = iris[,4])
> head(predict(rf, iris, type = "response"))
        1         2         3         4         5         6 
0.2519586 0.2026082 0.2093314 0.1946441 0.2344817 0.3564690 
> head(predict(rf, iris, type = "prob"))
Error in predict.randomForest(rf, iris, type = "prob") : 
  'prob' or 'vote' not meaningful for regression
```
@terrytangyuan @navdeep-G Maybe we can look around for some other uses of `type` in `predict()` in R and see if there's an other types/names we should be considering.

**Comment by terrytangyuan** [comment]
Thanks @ledell! 

I propose that we stick with the `type` argument for [caret's `predict()` method](https://www.rdocumentation.org/packages/caret/versions/5.05.004/topics/predict.train) since it's more unified across different R libraries. 

"A vector of predictions if type = "raw" or a data frame of class probabilities for type = "probs". In the latter case, there are columns for each class."

So options are: raw v.s. response and prob v.s. probs. Any preference? @ledell @navdeep-G 

As long as we have documentation for this I am okay with either one.

**Comment by terrytangyuan** [comment]
Regarding @ledell 's concern on probability:

"One concern I have about using predict_proba() underneath to produce probabilities is that it only works for sklearn models which is probably not what people will be using 95% of the time, and also requires much more computation. So my opinion would be not enable a type option which attempts to call either predict_proba() predict_log_proba() underneath."

@navdeep-G I am not too familiar with this yet. Is this true that it only works for sklearn models?

**Comment by ledell** [comment]
@terrytangyuan Yes, you'd have to use sklearn's implementation of [Platt scaling](http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/calibrate_model.html) to get the calibrated probabilities which are returned by `predict_proba()`, unless we wrote our own implementation inside of H2O4GPU (which is not on the roadmap right now).  Here's some more info about why the predictions are not really probabilities [here](http://danielnee.com/tag/platt-scaling/) and [here](http://www.datascienceassn.org/sites/default/files/Predicting%20good%20probabilities%20with%20supervised%20learning.pdf).

Using the term "probs" or "probabilities" is a misnomer (in my opinion) if you're not applying some sort of calibration to make sure they actually represent calibrated/posterior probabilities.  Scikit-learn uses the term probabilities correctly in the `predict_proba()` function, but when the `caret::predict.train()` method uses the term "probs" to return a data frame of "class probabilities", this seems incorrect to me since these are uncalibrated predicted values for each class.  This makes me hesitant to borrow this terminology. 

One way to skirt around the terminology issues is to return the same type of output that we do in H2O, a frame that contains both the predict class label and the predicted values for each class (no claim of actual probabilities, returning everything you might need at once).

I'm open to other ideas too.

**Comment by terrytangyuan** [comment]
I am okay with returning everything. My only concern about returning everything is that this will involve two for-loops (or one for-loop plus a [vector operation](https://github.com/h2oai/h2o4gpu/blob/master/src/interface_py/h2o4gpu/solvers/xgboost.py#L164-L165)) underneath where sometimes user don't want everything but they care about performance during inference time. It's not too much of an overhead but just want you to be aware of during the discussion. If in some cases, this list expands beyond just raw and probabilities, the overhead becomes larger.

**Comment by navdeep-G** [comment]
Is this still an issue?

**Comment by terrytangyuan** [comment]
Right now only `predict()` is exposed to R with the same behavior as Python API's `predict()`. I don't think it's a priority though.

**Comment by ledell** [comment]
I am going to close this, as it's been resolved in this PR: https://github.com/h2oai/h2o4gpu/pull/489
